Rewrite ShardedTensor.gather to use dist.gather instead of gather_object

kumpera opened the pull request:
gather_object is problematic when used with Tensors as they can unpickle on the wrong
device and lead to deadlocks or spurious failures.

This change introduces a RPC workaround for EFA when initing TensorPipe until
they properly address it.

Fixes #73935

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -918,7 +919,7 @@ def test_invalid_pg_rpc_ranks(self):
 
         # Init RPC with different ranks.
-        rpc_backend_options = rpc.TensorPipeRpcBackendOptions()
+        rpc_backend_options = rpc.TensorPipeRpcBackendOptions(_transports=tp_transports())
```
Do the EFA workarounds have anything to do with `ShardedTensor.gather`? If not, can we move these changes to a separate PR?

kumpera replied:
Sure, it's here because I have a hard time working on it on AWS otherwise.

pritamdamania87 replied:
Looks like we still have RPC changes here, shouldn't this go in https://github.com/pytorch/pytorch/pull/77363?

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -273,46 +278,85 @@ def gather(
                 Default: ``None``
         """
+        def shard_size(shard_md):
+            res = 1
+            for s in shard_md.shard_sizes:
+                res *= s
+            return res
```
Can't we just do `math.prod(shard_md.shard_sizes)`?

kumpera replied:
TIL

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -273,46 +278,85 @@ def gather(
+            if len(local_shards) > 0:
+                device = local_shards[0].tensor.device
+            else:
+                device = f"cuda:{torch.cuda.current_device()}"
```
Why don't we always use the current device? We do expect the user to set this appropriately as documented here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_shard/sharded_tensor/api.py#L109

kumpera replied:
Always using CUDA is wrong if this is a CPU ShardedTensor.

What we want is to alloc on `self._process_group` device which should be `cuda.current_device` if we're using NCCL or cpu with GLOO.

What I'm thinking here, and I need to add a test for, is how to handle `sharded_tensor.cpu().gather()` when there's one rank without data.

kumpera replied:
I replaced this with adding a method `_get_preferred_device` to ST which returns cuda:current_device for NCCL or CPU otherwise.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -273,46 +278,85 @@ def gather(
+                device = f"cuda:{torch.cuda.current_device()}"
+            data = torch.empty(max_rank_size, device=device)
+            for shard in local_shards:
+                for placement in local_shards_placement:
```
Why do we need `local_shards_placement`? Doesn't `local_shards` have everything you need?

kumpera replied:
This is a badly named variable, it tracks each shard offset in the exchange tensor we're building.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -273,46 +278,85 @@ def gather(
-                    out_narrow_view.copy_(tensor)
+        if rank != dst:
+            return
+        if out is None:
+            raise ValueError("`out` Tensor must be provided on dst rank!")
```
Probably do this validation right at the top of the function, not after doing expensive collectives.

kumpera replied:
This check is already done by _validate_output_tensor_for_gather but mypy can't see it. 
Replacing it with a cast and a comment on why.

pritamdamania87 reviewed (commented):
Thanks for fixing this!

rohan-varma commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -273,46 +278,85 @@ def gather(
+            if len(local_shards) > 0:
+                device = local_shards[0].tensor.device
+            else:
+                device = f"cuda:{torch.cuda.current_device()}"
```
nit: I don't think the device selection needs to be under no_grad context manager?

kumpera replied:
good point

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -35,6 +35,7 @@
     load_tests,
     sandcastle_skip_if,
     get_cycles_per_ms,
+    tp_transports,
```
Can we separate out the RPC changes to a different PR?

kumpera replied:
done

kumpera committed:
Switch ShardedTensor.gather to use dist.gather.

gather_object is problematic when used with Tensors as they can unpickle on the wrong
device and lead to deadlocks or spurious failures.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -273,46 +288,76 @@ def gather(
+        with torch.no_grad():
+            data = torch.empty(max_rank_size, device=self._get_preferred_device())
+            for shard in local_shards:
+                for placement in local_shards_offset:
```
nit: Might be cleaner if `local_shards_offsets` is a dict keyed on `ShardMetadata`, since `ShardMetadata` is a dataclass it probably already has `__eq__` and `__hash__` to enable this, if not maybe we can add it?

kumpera replied:
I had to make ShardMetadata and _remote_device hashable.
Which those changes, the `gather` code became significantly simpler.
Good catch!

kumpera replied:
I made that change, had to make both ShardMetadata and _remote_device  hashable.
The code turned out to be significantly clearer.

kumpera committed:
Make ShardMetadata and _remote_device hashable. Simplify gather.

kumpera commented:
@pytorchmergebot merge this
